### ISSUE TITLE
Spring cloud config watcher with Spring boot 3.x. Pods are not getting refreshed.Please help.

### ISSUE THREAD

**Issue by shashanksdixit** [spring-cloud/spring-cloud-kubernetes#1668]
Hi

I am trying to implement and test Spring Cloud Kubernetes Config Watcher. 
I have deployed the watcher as below. Nothing much is changed. The namespace is default.

<details><summary>Kube watcher yaml</summary>
<p>

```
---
apiVersion: v1
kind: List
items:
  - apiVersion: v1
    kind: Service
    metadata:
      labels:
        app: spring-cloud-kubernetes-configuration-watcher
      name: spring-cloud-kubernetes-configuration-watcher
    spec:
      ports:
        - name: http
          port: 8888
          targetPort: 8888
      selector:
        app: spring-cloud-kubernetes-configuration-watcher
      type: ClusterIP
  - apiVersion: v1
    kind: ServiceAccount
    metadata:
      labels:
        app: spring-cloud-kubernetes-configuration-watcher
      name: spring-cloud-kubernetes-configuration-watcher
  - apiVersion: rbac.authorization.k8s.io/v1
    kind: RoleBinding
    metadata:
      labels:
        app: spring-cloud-kubernetes-configuration-watcher
      name: spring-cloud-kubernetes-configuration-watcher:view
    roleRef:
      kind: Role
      apiGroup: rbac.authorization.k8s.io
      name: namespace-reader
    subjects:
      - kind: ServiceAccount
        name: spring-cloud-kubernetes-configuration-watcher
  - apiVersion: rbac.authorization.k8s.io/v1
    kind: Role
    metadata:
      namespace: default
      name: namespace-reader
    rules:
      - apiGroups: ["", "extensions", "apps"]
        resources: ["configmaps", "pods", "services", "endpoints", "secrets"]
        verbs: ["get", "list", "watch"]
  - apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: spring-cloud-kubernetes-configuration-watcher-deployment
    spec:
      selector:
        matchLabels:
          app: spring-cloud-kubernetes-configuration-watcher
      template:
        metadata:
          labels:
            app: spring-cloud-kubernetes-configuration-watcher
        spec:
          serviceAccount: spring-cloud-kubernetes-configuration-watcher
          containers:
          - name: spring-cloud-kubernetes-configuration-watcher
            image: springcloud/spring-cloud-kubernetes-configuration-watcher:3.1.2
            imagePullPolicy: IfNotPresent
            env: 
            - name: LOGGING_LEVEL_ORG_SPRINGFRAMEWORK_CLOUD_KUBERNETES_CONFIGURATION_WATCHER
              value: DEBUG
            - name: LOGGING_LEVEL_ORG_SPRINGFRAMEWORK_CLOUD_KUBERNETES_CLIENT_CONFIG_RELOAD
              value: DEBUG
            - name: LOGGING_LEVEL_ORG_SPRINGFRAMEWORK_CLOUD_KUBERNETES_COMMONS_CONFIG_RELOAD
              value: DEBUG
            - name: SPRING_CLOUD_KUBERNETES_CONFIGURATION_WATCHER_REFRESHDELAY
              value: "10000"
            readinessProbe:
              httpGet:
                port: 8888
                path: /actuator/health/readiness
            livenessProbe:
              httpGet:
                port: 8888
                path: /actuator/health/liveness
            ports:
            - containerPort: 8888

```
</p>
</details> 

I have configurations in properties file as below. I am making a config map from this file.

<details><summary>Properties file - application-kubernetes.properties</summary>
<p>

```
spring.application.name=logconfig

spring.cloud.kubernetes.reload.enabled=true
spring.cloud.kubernetes.reload.mode=event
spring.cloud.kubernetes.repload.monitoring-config-maps=true
spring.cloud.kubernetes.enabled=true
spring.cloud.kubernetes.config.enabled=true

spring.cloud.kubernetes.configuration.watcher.actuatorPort=8080
spring.config.import=optional:configserver:

management.endpoints.web.exposure.include=*
my.custom.property=InitialValueKubernetes1

logging.level.root=INFO
logging.level.io.sd=DEBUG

```

</p>
</details> 

I am testing currently the property change "my.custom.property". But it could be any property change though.
I have a configmap deployed as below. 

<details><summary>Configmap commands</summary>
<p>

```
kubectl create configmap my-config --from-file="/mnt/d/STSWorkspace/KubeConfigLogs/logconfig/src/main/resources/application-kubernetes.properties"
kubectl label  configmap my-config spring.cloud.kubernetes.config='true'
kubectl annotate configmap my-config spring.cloud.kubernetes.configmap.apps='logconfig'

```

</p>
</details> 

My deployment of the application looks as below.

<details><summary>Deployment Yml</summary>
<p>

```

apiVersion: apps/v1
kind: Deployment
metadata:
  name: logconfig
spec:
  replicas: 3
  selector:
    matchLabels:
      app: logconfig
  template:
    metadata:
      labels:
        app: logconfig
      annotations:
        configmap.reloader.stakater.com/reload: "my-config"
    spec:
      containers:
        - name: logconfig
          image: shashanksdixit/logconfig
          ports:
            - containerPort: 8080
          volumeMounts:
            - name: config-volume
              mountPath: /config
      volumes:
        - name: config-volume
          configMap:
            name: my-config


``` 

</p>
</details> 

When I am updating the configmap, and changing the value of "my.custom.property" to any other value, the changes are not refreshed.
I can see the changes in the /config/application-kubernetes.properties. But /refresh endpoint is never called. Here is the watcher log as below

<details><summary>Watcher Log</summary>
<p>

``` 
2024-06-24T08:52:46.339Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] sClientEventBasedConfigMapChangeDetector : ConfigMap my-config was updated in namespace default
2024-06-24T08:52:46.339Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-15-thread-1] sClientEventBasedConfigMapChangeDetector : ConfigMap my-config was updated in namespace default
2024-06-24T08:52:46.346Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : onEvent config-map: class V1ConfigMap {
    apiVersion: v1
    binaryData: null
    data: {application-kubernetes.properties=spring.application.name=logconfig

    spring.cloud.kubernetes.reload.enabled=true
    spring.cloud.kubernetes.reload.mode=event
    spring.cloud.kubernetes.repload.monitoring-config-maps=true
    spring.cloud.kubernetes.enabled=true
    spring.cloud.kubernetes.config.enabled=true

    spring.cloud.kubernetes.configuration.watcher.actuatorPort=8080
    spring.config.import=optional:configserver:
    #spring.config.import=kubernetes:,optional:file:/etc/config/application-kubernetes.properties
    #spring.config.import=configserver:kubernetes:

    management.endpoints.web.exposure.include=*
    my.custom.property=InitialValueKubernetes3

    logging.level.root=INFO
    logging.level.io.sd=DEBUG
    }
    immutable: null
    kind: ConfigMap
    metadata: class V1ObjectMeta {
        annotations: {spring.cloud.kubernetes.configmap.apps=logconfig}
        creationTimestamp: 2024-06-24T07:39:37Z
        deletionGracePeriodSeconds: null
        deletionTimestamp: null
        finalizers: null
        generateName: null
        generation: null
        labels: {spring.cloud.kubernetes.config=true}
        managedFields: null
        name: my-config
        namespace: default
        ownerReferences: null
        resourceVersion: 90713
        selfLink: /api/v1/namespaces/default/configmaps/my-config
        uid: f9c69c59-4f46-460b-95d5-b0691d5ccc16
    }
}
2024-06-24T08:52:46.353Z  INFO 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-15-thread-1] o.s.c.k.c.watcher.WatcherUtil            : will schedule remote refresh based on apps : [logconfig]
2024-06-24T08:52:46.360Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-15-thread-1] o.s.c.k.c.watcher.WatcherUtil            : Scheduling remote refresh event to be published for config-map: with appName : logconfig to be published in 10000 milliseconds
2024-06-24T08:52:46.441Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : environment from locateMapPropertySources : ApplicationReactiveWebEnvironment {activeProfiles=[kubernetes], defaultProfiles=[default], propertySources=[MapPropertySource {name='server.ports'}, ConfigurationPropertySourcesPropertySource {name='configurationProperties'}, PropertiesPropertySource {name='systemProperties'}, OriginAwareSystemEnvironmentPropertySource {name='systemEnvironment'}, RandomValuePropertySource {name='random'}, CachedRandomPropertySource {name='cachedrandom'}, MapPropertySource {name='KUBERNETES_NAMESPACE_PROPERTY_SOURCE'}, CompositePropertySource {name='composite-configmap', propertySources=[KubernetesClientConfigMapPropertySource {name='configmap.spring-cloud-kubernetes-configuration-watcher.default'}]}, CompositePropertySource {name='composite-secrets', propertySources=[]}, OriginTrackedMapPropertySource {name='Config resource 'class path resource [application.yml]' via location 'optional:classpath:/''}, MapPropertySource {name='springCloudClientHostInfo'}, MapPropertySource {name='spring.integration.poller'}, MapPropertySource {name='kafkaBinderDefaultProperties'},  {name='Management Server'}]}
2024-06-24T08:52:46.442Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : sources from locateMapPropertySources : [KubernetesClientConfigMapPropertySource {name='configmap.spring-cloud-kubernetes-configuration-watcher.default'}]
2024-06-24T08:52:46.445Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : environment from findPropertySources: ApplicationReactiveWebEnvironment {activeProfiles=[kubernetes], defaultProfiles=[default], propertySources=[MapPropertySource {name='server.ports'}, ConfigurationPropertySourcesPropertySource {name='configurationProperties'}, PropertiesPropertySource {name='systemProperties'}, OriginAwareSystemEnvironmentPropertySource {name='systemEnvironment'}, RandomValuePropertySource {name='random'}, CachedRandomPropertySource {name='cachedrandom'}, MapPropertySource {name='KUBERNETES_NAMESPACE_PROPERTY_SOURCE'}, CompositePropertySource {name='composite-configmap', propertySources=[KubernetesClientConfigMapPropertySource {name='configmap.spring-cloud-kubernetes-configuration-watcher.default'}]}, CompositePropertySource {name='composite-secrets', propertySources=[]}, OriginTrackedMapPropertySource {name='Config resource 'class path resource [application.yml]' via location 'optional:classpath:/''}, MapPropertySource {name='springCloudClientHostInfo'}, MapPropertySource {name='spring.integration.poller'}, MapPropertySource {name='kafkaBinderDefaultProperties'},  {name='Management Server'}]}
2024-06-24T08:52:46.447Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : environment sources from findPropertySources : [MapPropertySource {name='server.ports'}, ConfigurationPropertySourcesPropertySource {name='configurationProperties'}, PropertiesPropertySource {name='systemProperties'}, OriginAwareSystemEnvironmentPropertySource {name='systemEnvironment'}, RandomValuePropertySource {name='random'}, CachedRandomPropertySource {name='cachedrandom'}, MapPropertySource {name='KUBERNETES_NAMESPACE_PROPERTY_SOURCE'}, CompositePropertySource {name='composite-configmap', propertySources=[KubernetesClientConfigMapPropertySource {name='configmap.spring-cloud-kubernetes-configuration-watcher.default'}]}, CompositePropertySource {name='composite-secrets', propertySources=[]}, OriginTrackedMapPropertySource {name='Config resource 'class path resource [application.yml]' via location 'optional:classpath:/''}, MapPropertySource {name='springCloudClientHostInfo'}, MapPropertySource {name='spring.integration.poller'}, MapPropertySource {name='kafkaBinderDefaultProperties'},  {name='Management Server'}]
2024-06-24T08:52:46.449Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : findPropertySources : [configmap.spring-cloud-kubernetes-configuration-watcher.default]
2024-06-24T08:52:46.449Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : no changes found, reload will not happen
2024-06-24T08:52:46.451Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : No change detected in config maps/secrets, reload will not happen
2024-06-24T08:54:21.752Z  INFO 1 --- [spring-cloud-kubernetes-configuration-watcher] [s.V1Endpoints-1] i.k.c.informer.cache.ReflectorRunnable   : ResourceVersion 88566 and Watch connection expired: too old resource version: 88566 (89167) , will retry w/o resourceVersion next time
2024-06-24T08:54:21.753Z  INFO 1 --- [spring-cloud-kubernetes-configuration-watcher] [s.V1Endpoints-1] i.k.c.informer.cache.ReflectorRunnable   : class io.kubernetes.client.openapi.models.V1Endpoints#Watch expired, will re-list-watch soon
2024-06-24T08:54:22.754Z  INFO 1 --- [spring-cloud-kubernetes-configuration-watcher] [s.V1Endpoints-1] i.k.c.informer.cache.ReflectorRunnable   : class io.kubernetes.client.openapi.models.V1Endpoints#Start listing and watching...

``` 

</p>
</details> 

Please note that, the configmap changes are actually picked up. If you notice "my.custom.property" has its changed value as InitialValueKubernetes3 from InitialValueKubernetes1

Also, log says, 

```
2024-06-24T08:52:46.353Z  INFO 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-15-thread-1] o.s.c.k.c.watcher.WatcherUtil            : will schedule remote refresh based on apps : [logconfig]
2024-06-24T08:52:46.360Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-15-thread-1] o.s.c.k.c.watcher.WatcherUtil            : Scheduling remote refresh event to be published for config-map: with appName : logconfig to be published in 10000 milliseconds

``` 

But call to /refresh endpoint never happens.
Am I missing anything? Please guide.

You can see the full code at https://github.com/shashanksdixit/configwatcher/tree/main

**Comment by ryanjbaxter** [comment]
Can you please retry letting Spring Cloud manage the versions of its dependencies for you?

```
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>org.springframework.cloud</groupId>
        <artifactId>spring-cloud-dependencies</artifactId>
        <version>${spring-cloud.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
```

Can you also try specifying the namespace in which the configmap resides by adding this to your config watcher deployment
```env:
  - name: SPRING_CLOUD_KUBERNETES_RELOAD_NAMESPACES_0
    value: "namespace-a"
```

**Comment by shashanksdixit** [comment]
I believe I am already doing dependencies version management through <dependencyManagement>. Please let me know if I have missed anything. 

I have changed the watcher deployment as suggested. Now it looks like this (only part of the required section posted). 
```
env: 
            - name: LOGGING_LEVEL_ORG_SPRINGFRAMEWORK_CLOUD_KUBERNETES_CONFIGURATION_WATCHER
              value: DEBUG
            - name: LOGGING_LEVEL_ORG_SPRINGFRAMEWORK_CLOUD_KUBERNETES_CLIENT_CONFIG_RELOAD
              value: DEBUG
            - name: LOGGING_LEVEL_ORG_SPRINGFRAMEWORK_CLOUD_KUBERNETES_COMMONS_CONFIG_RELOAD
              value: DEBUG
            - name: SPRING_CLOUD_KUBERNETES_CONFIGURATION_WATCHER_REFRESHDELAY
              value: "10000"
            - name: SPRING_CLOUD_KUBERNETES_RELOAD_NAMESPACES_0
              value: "default"
``` 

But still, when I do GET /api/config, I only see the previous value of my.custom.property

**Comment by shashanksdixit** [comment]
I am testing. I will get back to you today itself.

**Comment by shashanksdixit** [comment]
Hi @wind57 

Thank you for clearning pom and updating code. 
I am using same deployment files and service. But my result is as earlier. No change detected. :(
 The only difference is I am using service type as NodePort. I am testing in my local with KinD on WSL2. I was not able to create service type as LoadBalancer, so I created as NodePort.  Rest everything is same.

**Comment by shashanksdixit** [comment]
> Hi @wind57
> 
> Thank you for clearning pom and updating code. I am using same deployment files and service. But my result is as earlier. No change detected. :( The only difference is I am using service type as NodePort. I am testing in my local with KinD on WSL2. I was not able to create service type as LoadBalancer, so I created as NodePort. Rest everything is same.
> 
> Edit - If I call the /actuator/refresh manually using service endpoint, then I see the value changing. I am not sure why config watcher is not doing it.

Logs for reference -
I tested the following cases.
1. I created the configmap from configmap.yml and changed the my.custom.property by editing via "kubectl edit configmap"
2. I created the configmap from application-kubernetes.properties and changed the my.custom.property by editing via "kubectl edit configmap"
3. I created the configmap from application-kubernetes.properties, deleted the configmap, recreated it from configmap.yml

In all the cases, the Config-watcher picks up the change (as in it logs the recent change), but it also fails to recognize the change. Below are the common logs in all the above scenarios.
```
2024-06-28T04:19:10.113Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-15-thread-1] o.s.c.k.c.watcher.WatcherUtil            : Scheduling remote refresh event to be published for config-map: with appName : logconfig to be published in 30000 milliseconds
2024-06-28T04:19:10.123Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : environment from locateMapPropertySources : ApplicationReactiveWebEnvironment {activeProfiles=[kubernetes], defaultProfiles=[default], propertySources=[MapPropertySource {name='server.ports'}, ConfigurationPropertySourcesPropertySource {name='configurationProperties'}, PropertiesPropertySource {name='systemProperties'}, OriginAwareSystemEnvironmentPropertySource {name='systemEnvironment'}, RandomValuePropertySource {name='random'}, CachedRandomPropertySource {name='cachedrandom'}, MapPropertySource {name='KUBERNETES_NAMESPACE_PROPERTY_SOURCE'}, CompositePropertySource {name='composite-configmap', propertySources=[KubernetesClientConfigMapPropertySource {name='configmap.spring-cloud-kubernetes-configuration-watcher.default'}]}, CompositePropertySource {name='composite-secrets', propertySources=[]}, OriginTrackedMapPropertySource {name='Config resource 'class path resource [application.yml]' via location 'optional:classpath:/''}, MapPropertySource {name='springCloudClientHostInfo'}, MapPropertySource {name='spring.integration.poller'}, MapPropertySource {name='kafkaBinderDefaultProperties'},  {name='Management Server'}]}
2024-06-28T04:19:10.123Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : sources from locateMapPropertySources : [KubernetesClientConfigMapPropertySource {name='configmap.spring-cloud-kubernetes-configuration-watcher.default'}]
2024-06-28T04:19:10.123Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : environment from findPropertySources: ApplicationReactiveWebEnvironment {activeProfiles=[kubernetes], defaultProfiles=[default], propertySources=[MapPropertySource {name='server.ports'}, ConfigurationPropertySourcesPropertySource {name='configurationProperties'}, PropertiesPropertySource {name='systemProperties'}, OriginAwareSystemEnvironmentPropertySource {name='systemEnvironment'}, RandomValuePropertySource {name='random'}, CachedRandomPropertySource {name='cachedrandom'}, MapPropertySource {name='KUBERNETES_NAMESPACE_PROPERTY_SOURCE'}, CompositePropertySource {name='composite-configmap', propertySources=[KubernetesClientConfigMapPropertySource {name='configmap.spring-cloud-kubernetes-configuration-watcher.default'}]}, CompositePropertySource {name='composite-secrets', propertySources=[]}, OriginTrackedMapPropertySource {name='Config resource 'class path resource [application.yml]' via location 'optional:classpath:/''}, MapPropertySource {name='springCloudClientHostInfo'}, MapPropertySource {name='spring.integration.poller'}, MapPropertySource {name='kafkaBinderDefaultProperties'},  {name='Management Server'}]}
2024-06-28T04:19:10.123Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : environment sources from findPropertySources : [MapPropertySource {name='server.ports'}, ConfigurationPropertySourcesPropertySource {name='configurationProperties'}, PropertiesPropertySource {name='systemProperties'}, OriginAwareSystemEnvironmentPropertySource {name='systemEnvironment'}, RandomValuePropertySource {name='random'}, CachedRandomPropertySource {name='cachedrandom'}, MapPropertySource {name='KUBERNETES_NAMESPACE_PROPERTY_SOURCE'}, CompositePropertySource {name='composite-configmap', propertySources=[KubernetesClientConfigMapPropertySource {name='configmap.spring-cloud-kubernetes-configuration-watcher.default'}]}, CompositePropertySource {name='composite-secrets', propertySources=[]}, OriginTrackedMapPropertySource {name='Config resource 'class path resource [application.yml]' via location 'optional:classpath:/''}, MapPropertySource {name='springCloudClientHostInfo'}, MapPropertySource {name='spring.integration.poller'}, MapPropertySource {name='kafkaBinderDefaultProperties'},  {name='Management Server'}]
2024-06-28T04:19:10.124Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : findPropertySources : [configmap.spring-cloud-kubernetes-configuration-watcher.default]
2024-06-28T04:19:10.124Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : no changes found, reload will not happen
2024-06-28T04:19:10.124Z DEBUG 1 --- [spring-cloud-kubernetes-configuration-watcher] [ool-11-thread-1] o.s.c.k.c.c.reload.ConfigReloadUtil      : No change detected in config maps/secrets, reload will not happen

```

**Comment by shashanksdixit** [comment]
This works.


First and foremost, I would like to extend my gratitude to @wind57 for your assistance and the updates you provided. Thank you also for your patience with me :) Additionally, I would like to thank @ryanjbaxter for your prompt responses and help.

Two things I did at my side to get the consistent result.

Earlier when the configmap was modified by "kubectl edit", the end result was not consistent i.e. sometimes I see updates at the endpoint and sometimes I did not. Then I increased the watcher refresh delay to 2 mins. This has resulted in consistent results and then I noticed that in the watcher log that I am getting the logs where watcher is calling the /refresh endpoint.

I also attempted to create the configmap from the command line using "kubectl create" with the "from file" option. After creation, I would label and annotate it appropriately. However, since, technically it was modified before labeling, it didn't seem to work. Now, I am creating the configmap using a YAML file, copying all the properties into the configmap YAML to test if it works for all properties as well.

Thanks a lot again.
